### PR TITLE
Add Plural Support to paraglide

### PR DESCRIPTION
**Pull request by LorisSigrist**
This PR adds plural-support to paraglide using the new SDKv2. It uses `loadProjectFromDirectoryInMemory` under the hood.

There are still some edge-cases that are expressible in the AST but not supported in the compiler, however, those should be nearly impossible to run in to in real-world usage. 

The JSDoc comments aren't yet identical to the ones we had before.

Run the new compiler with `paraglide-js compile2 --project ./project.inlang`

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -59,5 +59,6 @@
 		"prettier": "2.8.3",
 		"typescript": "^5.5.2",
 		"vitest": "0.34.6"
-	}
+	},
+	"packageManager": "pnpm@9.3.0+sha512.ee7b93e0c2bd11409c6424f92b866f31d3ea1bef5fbe47d3c7500cdc3c9668833d2e55681ad66df5b640c61fa9dc25d546efa54d76d7f8bf54b13614ac293631"
```
Comment (samuelstroschein) [inline package.json:63]: Is this an accidental change?

File: inlang/source-code/sdk2/src/plugin/importPlugins.ts [hunk 2]
```diff
@@ -55,3 +55,19 @@ async function fetchModuleWithCache(uri: string): Promise<string> {
...
+		);
+	}
+	return await response.text();
+}
+
+/**
+ * Reads the module from disk
+ * @param path
+ */
+function fetchModuleFromPath(path: string): Promise<string> {}
```
Comment (samuelstroschein) [inline inlang/source-code/sdk2/src/plugin/importPlugins.ts:73]: Code doesn't seem to be used?

Comment (samuelstroschein) [reply inlang/source-code/sdk2/src/plugin/importPlugins.ts:73]: Unintended change? (Remove?)

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/aliases.ts [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+
+/**
+ * Returns re-export statements for each alias of a MessageBundle.
+ * If no aliases are present, this function returns an empty string.
+ *
+ * @param bundle
+ */
+export function reexportAliases(bundle: BundleNested) {
+	const aliases = Object.values(bundle.alias)
+	if (aliases.length > 1) throw new Error("Only one alias is allowed per bundle") // really?
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/aliases.ts:11]: No. You can compile all aliases that are valid js identifiers.

Comment (samuelstroschein) [reply inlang/source-code/paraglide/paraglide-js/src/compilerV2/aliases.ts:11]: The logic here should change. Compile every alias that is a valid JS identifier. The concept of a default alias has been removed. See https://github.com/opral/monorepo/pull/3048#discussion_r1707395555 for context

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/bundleIndex.ts [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	 */
+	availableLanguageTags: string[]
+}) => {
+	const hasInputs = Object.keys(args.typeRestrictions).length > 0
+
+	return `/**
+ * This message has been compiled by [inlang paraglide](https://inlang.com/m/gerre34r/library-inlang-paraglideJs).
+ *
+ * - Don't edit the message's code. Use [Sherlock (VS Code extension)](https://inlang.com/m/r7kp499g/app-inlang-ideExtension),
+ *   the [web editor](https://inlang.com/m/tdozzpar/app-inlang-finkLocalizationEditor) instead, or edit the translation files manually.
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/bundleIndex.ts:28]:
```suggestion
 *   [Fink](https://inlang.com/m/tdozzpar/app-inlang-finkLocalizationEditor) instead, or edit the translation files manually.
```

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compile.ts [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+	//Maps each language to it's fallback
+	//If there is no fallback, it will be undefined
+	const fallbackMap = getFallbackMap(opts.settings.locales, opts.settings.baseLocale)
+
+	const resources = opts.bundles.map((bundle) => compileBundle(bundle, fallbackMap))
+
+	const indexFile = [
+		"/* eslint-disable */",
+		'import { languageTag } from "./runtime.js"',
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compile.ts:52]: Not for this PR but to keep in mind for paraglide v2, deprecate `languageTag` in favor of `locale` (compile both. v3 removes `languageTag`)

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compile.ts [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+		opts.settings.locales
+			.map((locale) => `import * as ${i(locale)} from "./messages/${locale}.js"`)
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compile.ts:54]: I suggest renaming `i` to `jsIdentifier` or similar. 

I was wondering where `i` was coming from and what `i` is. An index? Why index over the locale? Using `jsIdentifier` would improve the comprehension.

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compile.ts [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+async function fmt(js: string): Promise<string> {
+	return await prettier.format(js, {
+		arrowParens: "always",
+		singleQuote: true,
+		parser: "babel",
+		plugins: ["prettier-plugin-jsdoc"],
+	})
+}
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compile.ts:126]: (No action required)

Such an easy workaround to get formatting right :+1. 

Besides, I find reading source code easier than an AST framework. I hope that we get around an AST. It makes contributions, reviews, etc. easier.

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compile.ts [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+	for (const locale of opts.settings.locales) {
+		const filename = `messages/${locale}.js`
+		let file = ["/* eslint-disable */", "import * as registry from '../registry.js' "].join("\n")
+
+		for (const resource of resources) {
+			// Bundle Aliases
+			const alias = resource.bundle.source.alias
+				? Object.values(resource.bundle.source.alias).at(0)
+				: undefined
+
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compile.ts:78]: Flaggin here that it should compile all valid aliases.

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.test.ts [hunk 9]
```diff
@@ -0,0 +1,159 @@
...
+
+		",
+		    "source": {
+		      "alias": {},
+		      "id": "my_bundle",
+		      "messages": [
+		        {
+		          "bundleId": "my_bundle",
+		          "declarations": [],
+		          "id": "c35d5203-874a-4c6c-82f2-e3e9938c79d6",
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.test.ts:56]: how do you get stable ids for testing?

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.test.ts [hunk 10]
```diff
@@ -0,0 +1,159 @@
...
+				text: "Hallo Welt!",
+			}),
+		],
+	})
+
+	const output = compileBundle(bundle, {
+		en: undefined,
+		de: "en",
+	})
+	expect(output).toMatchInlineSnapshot(`
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.test.ts:27]: Why did you move to snapshot testing instead of the rollup code testing setup we once had? 

- snapshot change every time JSdoc, ids, etc are updated, that seems like a lot of hassle
- calling the message functions and comparing the output seems much easier

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+/**
+ * Compiles all the messages in the bundle and returns an index-function + each compiled message
+ */
+export const compileBundle = (
+	bundle: BundleNested,
+	fallbackMap: Record<string, string | undefined>
+): Resource => {
+	if (!isValidJSIdentifier(bundle.id)) {
+		throw new Error(
+			`Cannot compile message with ID "${bundle.id}".\n\nThe message is not a valid JavaScript variable name. Please choose a different ID.\n\nTo detect this issue during linting, use the valid-js-identifier lint rule: https://inlang.com/m/teldgniy/messageLintRule-inlang-validJsIdentifier`
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.ts:25]:
```suggestion
			`Cannot compile bundle with ID "${bundle.id}".\n\nThe bundle id is not a valid JavaScript variable name. This is likely an internal error. An ID should always be valid. Open an issue in https://github.com/opral/monorepo. 
```

Comment (samuelstroschein) [reply inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.ts:25]: The bundle ids are provided by the SDK and should always be valid. This error shouldn't happen./

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileExpression.ts [hunk 12]
```diff
@@ -0,0 +1,51 @@
...
+	// TODO Type-Narrowing for options
+	return { code, typeRestrictions: {}, source: options }
+}
+
+function compileArg(arg: Expression["arg"]): string {
+	switch (arg.type) {
+		case "literal":
+			return `"${escapeForDoubleQuoteString(arg.name)}"`
+		case "variable": {
+			const escaped = !isValidJSIdentifier(arg.name)
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileExpression.ts:47]: Aha. Function, variable, input names can't be compiled if they are not valid JS identifiers. Should we enforce `snake_case` for Names in the inlang SDK? 

- `snake_case` is the most interoperable format 
- i fear that we will get into `my-cool-function` scenarios which can't be compiled in several programming languages, or `myCoolFunction` which requires mapping in other programming languages that are case insensitive (does that exist?)

Comment (samuelstroschein) [reply inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileExpression.ts:47]: We could have a validation rule for it though "use_snake_case_validation_rule"

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileMessage.test.ts [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+					type: "input",
+					name: "secondInput",
+					value: { type: "expression", arg: { type: "variable", name: "second Input" } },
+				},
+			],
+			selectors: [
+				{ type: "expression", arg: { type: "variable", name: "fistInput" } },
+				{
+					type: "expression",
+					arg: { type: "variable", name: "second Input" },
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileMessage.test.ts:27]: This can't work? The input name has whitespace which does not match the declarations. Hmmm supports the "we should enforce snake_case for names" argument. Do you agree?

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileMessage.test.ts [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+			"second Input": "number",
+		})
+		expect(compiled.code).toMatchInlineSnapshot(`
+			"/**
+			 * @param {{ 'second Input': number, fistInput: NonNullable<unknown> }} inputs
+			 * @returns {string}
+			 */
+			/* @__NO_SIDE_EFFECTS__ */
+			const some_message = (inputs) => {
+				const selectors = [ inputs.fistInput, registry.plural(\\"en\\", inputs['second Input']) ]
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileMessage.test.ts:58]: (no action needed)

beautiful comprehension. reading `registry.plural` immediately makes click "aha this function is coming from the registry".

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/inputsType.ts [hunk 15]
```diff
@@ -0,0 +1,30 @@
...
+
+export type InputTypeMap = Record<string, string>
+
+/**
+ * Generates a JSDoc comment for an input object
+ *
+ * Returns an empty string if the params object is empty.
+ *
+ * @example
+ *   const jsdoc = jsdocFromParams({ name: "NonNullable<unknown>", count: "NonNullable<unknown>" })
```
Comment (samuelstroschein) [inline inlang/source-code/paraglide/paraglide-js/src/compilerV2/inputsType.ts:12]: Outdated JSDoc?

**Review by samuelstroschein — CHANGES_REQUESTED** [review]
Is OK. Left comments. Change that should land before approval: 

- All valid js identifier bundle aliases should be compiled

File: package.json (hunk 1 above)
Comment (LorisSigrist) [reply package.json:63]: Yes, will revert

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileExpression.ts (hunk 12 above)
Comment (LorisSigrist) [reply inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileExpression.ts:47]: Paraglide itself could work with any arbitrary string for:
- IDs & Aliases
- Registry Function names & arguments
- Message Inputs

From Paraglide's perspective we don't need to enforce any specific formatting. I don't know about other Inlang apps though

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.ts (hunk 11 above)
Comment (LorisSigrist) [reply inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.ts:25]: I'll remove the failure case

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileMessage.test.ts (hunk 13 above)
Comment (LorisSigrist) [reply inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileMessage.test.ts:27]: Having spaces in variable names is perfectly ok. The space in the test is on purpose to make sure it works as expected

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.test.ts (hunk 10 above)
Comment (LorisSigrist) [reply inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileBundle.test.ts]: I switched back to the bundling based tests. 

I prefer snapshot tests for developing since they do a good job at showing the output. The bundling based tests are more robust once you have something that works

File: inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileExpression.ts (hunk 12 above)
Comment (samuelstroschein) [reply inlang/source-code/paraglide/paraglide-js/src/compilerV2/compileExpression.ts:50]: Interesting. Then let's skip the enforcement

**Review by samuelstroschein — APPROVED** [review]

**Comment by samuelstroschein** [comment]
@LorisSigrist merge?
